src/meta/greedy_load_balancer.cpp

```diff
@@ -963,10 +1000,120 @@ bool greedy_load_balancer::cluster_replica_balance(const cluster_balance_type ty
     return true;
 }
 
-bool greedy_load_balancer::do_cluster_replica_balance(const cluster_balance_type type)
+bool greedy_load_balancer::do_cluster_replica_balance(const meta_view *global_view,
+                                                      const cluster_balance_type type)
 {
-    return true;
+    cluster_migration_info cluster_info;
+    if (!get_cluster_migration_info(global_view, type, cluster_info)) {
+        return false;
+    }
+
     /// TBD(zlw)
+    return true;
+}
+
+bool greedy_load_balancer::get_cluster_migration_info(const meta_view *global_view,
+                                                      const cluster_balance_type type,
+                                                      /*out*/ cluster_migration_info &cluster_info)
+{
+    const node_mapper &nodes = *global_view->nodes;
+    if (nodes.size() < 3) {
+        return false;
+    }
+    for (const auto &kv : nodes) {
+        if (!all_replica_infos_collected(kv.second)) {
+            return false;
+        }
+    }
+
+    const app_mapper &all_apps = *global_view->apps;
+    app_mapper apps;
+    for (const auto &kv : all_apps) {
+        const std::shared_ptr<app_state> &app = kv.second;
+        if (is_ignored_app(app->app_id) || app->is_bulk_loading || app->splitting()) {
+            return false;
```

Comment:
why meet `ignored_app` is `return`? actually should `continue`?

Reply:
yes, you are right